contrib/analysis/aggregate_data.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+
+import argparse
+
+# import multiprocessing
+import requests
+import time
+
+from datetime import datetime
+from dateutil import rrule
+from dateutil.relativedelta import relativedelta
+from pbench_combined_data import PbenchCombinedDataCollection
+
+from elasticsearch1 import Elasticsearch
+
+
+def _month_gen(now: datetime) -> str:
+    """Generate YYYY-MM stings from all the months, inclusively, between the
+    given start and end dates.
+    """
+    start = now - relativedelta(years=1)
+    first_month = start.replace(day=1)
+    last_month = now + relativedelta(day=31)
+    for m in rrule.rrule(rrule.MONTHLY, dtstart=first_month, until=last_month):
+        yield f"{m.year:04}-{m.month:02}"
+
+
+def main(args):
+
+    incoming_url = f"{args.url_prefix}/incoming/"
+
+    if args.profile_memory_usage:
+        from guppy import hpy
+
+        memprof = hpy()
+    else:
+        memprof = None
+
+    if memprof:
+        print(f"Initial memory profile ... {memprof.heap()}", flush=True)
+
+    # We create the multiprocessing pool first to avoid forking a sub-process
+    # with lots of memory allocated.
+    # ncpus = multiprocessing.cpu_count() - 1 if args.cpu_n == 0 else args.cpu_n
+    # pool = multiprocessing.Pool(ncpus) if ncpus != 1 else None
+
+    es = Elasticsearch(
+        [f"{args.es_host}:{args.es_port}"], timeout=60
+    )  # to prevent read timeout errors (60 is arbitrary)
+
+    session = requests.Session()
+    ua = session.headers["User-Agent"]
+    session.headers.update({"User-Agent": f"{ua} -- merge_sos_and_perf_parallel"})
+    pbench_data = PbenchCombinedDataCollection(
+        incoming_url, session, es, args.record_limit
+    )
+
+    scan_start = time.time()
+    now = datetime.utcfromtimestamp(scan_start)
+
+    # TODO: This doesn't work because es instance not pickle-able.
+    # pool.starmap(merge_run_result_index, [(es, month, args.record_limit, pbench_data) for month in _month_gen(now)])
+
+    for month in _month_gen(now):
+        pbench_data.collect_data(month)
+
+    # NOTE: Not writing sosreports and results to files. Will work on this step
+    #       of sosreport processing, etc next.
+
+    scan_end = time.time()
+    duration = scan_end - scan_start
+
+    print(pbench_data)
```

Comment:
Perhaps add a "report" method to the `PbenchCombinedDataCollection` object, and another method of emitting the `.csv` files.

Reply:
Added print_report method to print tracker information to terminal, and emit_csv method that writes important data collected like valid,invalid,diskhost names, client names, and tracker info to separate csv files in a subdirectory. See new commit.